src/app/utils/state.ts

```diff
@@ -0,0 +1,123 @@
+import { SupabaseClient, createClient } from "@supabase/supabase-js";
+import { INTERACTION } from "./interaction";
+import { pipeline } from "@xenova/transformers";
+import { PromptTemplate } from "langchain/prompts";
+import { getModel } from "./model";
+import { LLMChain } from "langchain/chains";
+import { data } from "autoprefixer";
+
+const embedding_endpoint = process.env.SUPABASE_EMBEDDING_ENDPOINT!;
+
+class StateManager {
+  private static instance: StateManager;
+
+  private dbClient: SupabaseClient;
+  private model: any;
+
+  public constructor() {
+    const auth = {
+      detectSessionInUrl: false,
+      persistSession: false,
+      autoRefreshToken: false,
+    };
+    const url = process.env.SUPABASE_URL!;
+    const privateKey = process.env.SUPABASE_PRIVATE_KEY!;
+    this.dbClient = createClient(url, privateKey, { auth });
+    this.model = getModel();
+    this.model.verbose = true;
+  }
+
+  public async init() {}
+
+  public async update() {
```

Comment:
**issue (complexity):** Consider breaking down the update method into smaller, focused helper methods

The `update()` method can be simplified by extracting focused helper methods while maintaining the StateManager's coordination role. This improves readability and maintainability without over-engineering:

```typescript
class StateManager {
  // ... existing code ...

  private async getLatestStatus() {
    const { data: status } = await this.dbClient
      .from("tamagotchi_status")
      .select()
      .order("updatedat", { ascending: false })
      .limit(1);
    return status![0];
  }

  private async getRecentInteractions() {
    const { data: interactions } = await this.dbClient
      .from("tamagotchi_interactions")
      .select()
      .order("ts", { ascending: true })
      .limit(10);
    return interactions!;
  }

  private formatInteractions(interactions: any[]) {
    return interactions
      .map((interaction) => 
        `Interaction: ${INTERACTION[interaction.interaction]}, ${JSON.stringify(interaction.metadata)}, at ${interaction.ts}`
      )
      .join("\n ");
  }

  public async update() {
    const status = await this.getLatestStatus();
    const age = status.age ? status.age + 1 : 1;

    const interactions = await this.getRecentInteractions();
    const formattedInteractions = this.formatInteractions(interactions);

    const stateChain = new LLMChain({
      llm: this.model,
      prompt: PromptTemplate.fromTemplate(/* ... existing template ... */)
    });

    const result = await stateChain.call({
      lastInteractions: formattedInteractions,
      status: JSON.stringify({
        health: status.status.health,
        happiness: status.status.happiness,
        hunger: status.status.hunger,
      })
    });

    const newStatus = { ...JSON.parse(result!.text), age };
    await this.dbClient
      .from("tamagotchi_status")
      .insert({
        status: newStatus,
        updatedat: new Date().toISOString(),
      });
  }
}